pkg/window/strategy/session/session.go

```diff
@@ -0,0 +1,389 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package session
+
+import (
+	"math"
+	"strings"
+	"time"
+
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/reduce/pbq/partition"
+	"github.com/numaproj/numaflow/pkg/window"
+)
+
+var delimiter = ":"
+
+// Partition is a common partition for session window.
+// session windows share a common pbq, this partition is used to identify the pbq instance.
+var Partition = partition.ID{
+	Start: time.UnixMilli(0),
+	End:   time.UnixMilli(math.MaxInt64),
+	Slot:  "slot-0",
+}
+
+// sessionWindow TimedWindow implementation for Session window.
+type sessionWindow struct {
+	startTime time.Time
+	endTime   time.Time
+	slot      string
+	keys      []string
+}
+
+func NewSessionWindow(startTime time.Time, gap time.Duration, message *isb.ReadMessage) window.TimedWindow {
+	start := startTime
+	end := start.Add(gap)
+	// TODO: slot should be extracted based on the key
+	// we can accept an interface SlotAssigner
+	// which will assign the slot based on the keys
+	slot := "slot-0"
+	return &sessionWindow{
+		startTime: start,
+		endTime:   end,
+		slot:      slot,
+		keys:      message.Keys,
+	}
+}
+
+func (w *sessionWindow) StartTime() time.Time {
+	return w.startTime
+}
+
+func (w *sessionWindow) EndTime() time.Time {
+	return w.endTime
+}
+
+func (w *sessionWindow) Slot() string {
+	return w.slot
+}
+
+func (w *sessionWindow) Keys() []string {
+	return w.keys
+}
+
+func (w *sessionWindow) Partition() *partition.ID {
```

Comment:
```
// Partition returns the partition id of the window
	// which will be used to map to the pbq instance.
	Partition() *partition.ID
```
Should we use `SharedSessionPartition` here for the session window? Since it's used for identifying the PBQ.

Reply:
both are a little different, I will update and fix the comments to avoid confusion.